.github/workflows/latest_deps.yml

```diff
@@ -0,0 +1,160 @@
+# People who are freshly `pip install`ing from PyPI will pull in the latest versions of all
+# dependencies. Since most CI runs are against the locked poetry environment, run
+# specifically against the latest dependencies to know if there's an upcoming breaking
+# change.
+#
+# To plug that hole, this job is a canary which periodically
+# - checks out develop,
+# - pip installs from source (building a wheel locally using `poetry-core`,
+#   and `pip install`ing from that), and
+# - runs test suites against that checkout.
+#
+# Based on the twisted trunk CI job.
+
+name: Latest dependencies
+
+on:
+  schedule:
+    - cron: 0 7 * * *
+  pull_request:
+    branches: [dmr/poetry-wheel-ci]
+  push:
+    branches: [dmr/poetry-wheel-ci]
+  workflow_dispatch:
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  mypy:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      # The dev dependencies aren't exposed in the wheel metadata (at least with current
+      # poetry-core versions), so we install with poetry.
+      - uses: matrix-org/setup-python-poetry@v1
+        with:
+          python-version: "3.x"
+          poetry-version: "1.2.0b1"
+      # Dump installed versions for debugging.
+      - run: poetry run pip list > before.txt
+      # Upgrade all runtime dependencies only. This is intended to mimic a fresh
+      # `pip install matrix-synapse[all]` as closely as possible.
```

Comment:
"as closely as possible" pretty much means "except for the different dependency resolvers"?

Reply:
That's exactly right.

I spent a lot of time trying to get a vanilla pip install here, so that we wouldn't have to be "as close as possible". It was surprisingly tricky, and in the end I begrudingingly admitted that `poetry update` was probably good enough.

Reply:
Below this (for the trial tests) we use `pip install -e .` though, so I'm a bit confused why we can't do anything similar here?

I guess I'm confused at why we're doing it two different ways.

Reply:
I was just about to add a comment to that effect. `pip install -e .` only works because we're still using setuptools as the build system. (See the bottom of pyproject.toml). Poetry 1.1 doesn't suppose `-e` installs; I think it's planned to be supported for 1.2. (See also PEP660).

So I think that means that this will break once #12478 lands.

Reply:
> Below this (for the trial tests) we use `pip install -e .` though, so I'm a bit confused why we can't do anything similar here?

The tl;dr is that we need to install mypy, mypy-zope, etc. which is only known to poetry (you cannot install these via pip).

Reply:
(Whereas `trial` and `sytest` need no such dev-only dependencies.)